docs/design/proposals/creds-from-kms.md

```diff
@@ -0,0 +1,131 @@
+# Fetching Ceph user credentials from KMS
+
+## Proposal
+
+Ceph-CSI supports several KMS implementations for storing and
+fetching DEKs (Data Encryption Keys) used for RBD volume encryption.
+The focus of this proposal is to leverage existing KMS implementation
+to support obtaining Ceph user keyring used while
+creating, deleting, mapping and resizing volumes.
+
+### Benefits
+
+- For scenarios where Ceph is deployed externally, either as
+  a standalone or through Rook, one needs to obtain the Ceph keyring(s)
+  and manually create the K8s secrets needed by Ceph-CSI. With this feature,
+  sensitive keyring(s) can be stored securely in an
+  external secret management system such as Hashicorp Vault,
+  AWS Secret Manager, etc., and have Ceph-CSI pull these
+  keys directly from them.
+- More secure compared to K8s secrets,
+  which are relatively easy to access and
+  store sensitive secrets as unencrypted based64 encoded text.
+
+### Drawbacks
+
+- Adds additional overhead as on each CSI RPC call
+  a request is made to KMS for fetching credentials.
+- Possible risk of hitting the KMS rate limit.
+
+## Extending existing implementation
+
+- KMS ID is provided as part of the provisioner,
+  node-stage, and controller-expand secrets, along with Ceph `userID`, as shown below.
+  Having this information as part of *StorageClass* parameters,
+  like using the `encryptionKMSID` key for volume encryption,
+  wouldn't work as these parameters are not passed
+  to all CSI RPCs, e.g. `DeleteVolume`, `ControllerExpand` etc.
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: csi-rbd-provisioner
+stringData:
+  userID: <ceph-user-id>
+  kmsID: <kms-id>
+ ```
+
+- KMS config entry corresponding to `kmsID` needs to be added to existing
+  KMS ConfigMap, very similar to the volume encryption config found in
+  [cph-csi-encryption-kms-config](https://github.com/ceph/ceph-csi/blob/devel/examples/kms/vault/kms-config.yaml).
+
+- For KMS that rely on tenants' namespace to obtain required
+  ServiceAccount, ConfigMap, or Secrets,
+  an optional field, "tenantNamespace" will be added to the KMS config.
+  As the tenant namespace isn't available in all CSI requests,
+  value from provided in this field will be used instead.
+  If "tenantNamespace" field is absent, Ceph-CSI namespace
+  will be used as default.
```

Comment:
I think this is not very practical in the end. Why have a _default_ user for some operations, and use the tenant for others? Using different Ceph credentials for different operations on a single volume looks rather confusing. Usually you would also try to prevent normal users from accessing the Ceph cluster, so those users should not need to have access to Ceph credentials.

Reply:
@nixpanic to clarify my point:
1. the `tenantNamespace` here is only used for getting everything required by the KMS to fetch the Ceph user credentials. One such example is obtaining the KMS access token, which could be creating a service account token for vault SA or fetching tokens from K8s secret for some other KMS. 
2. Any volume associated with a storage class would always use same Ceph credentials for all its operations. I see Ceph credentials as pair of `userID and userKey`, where userID is obtained from `csi.storage.k8s.io/*-secret-name`  and `userKey` is fetched from the KMS. The user will never change until we change `userID` and KMS config. 
3. The KMS will only use one namespace for all the operations. If the `tenantNamespace` is present in the KMS config, it will always use that. If not, it always looks in Ceph-CSI namespace as default. Again, this is only for KMS authentication so that we can fetch the required Ceph credentials.